tests/model_integration_tests.py

```diff
@@ -102,13 +110,213 @@ def test_dummy_cache(self):
+        """
+        for i in range(5):
+            Car.objects.get(id=self.car.id).engine.name
+        # The above query is actually composed of 2 sql queries:
+        # one for fetching cars and the other engines.
+        self.assertEqual(len(connection.queries), 2)
+
+    @override_settings(DEBUG=True)
+    def test_one_to_one_mapping_cache_with_save(self):
+        """
+        Cache should be invalidated when calling 'save' on related objects
+        """
+        Car.objects.get(id=self.car.id).engine.name
+        self.engine.name = 'new_engine'
+        self.engine.save()
+        reset_queries()
+
+        # Only 1 cache (the one for engine selection query) will be invalidated
+        # as we only update data on Engine table
+        Car.objects.get(id=self.car.id).engine.name
+        self.assertEqual(len(connection.queries), 1)
+
+    @override_settings(DEBUG=True)
+    def test_one_to_one_mapping_cache_with_delete(self):
+        """
+        Cache should be invalidated when calling 'delete' related objects
+        """
+        Car.objects.get(id=self.car.id).engine.name
+        self.engine.delete()
+        reset_queries()
+
+        with self.assertRaises(ObjectDoesNotExist):
+            Car.objects.get(id=self.car.id).engine.name
+
+    @override_settings(DEBUG=True)
+    def test_many_to_many_mapping_cache(self):
+        """
+        Queries should be cached when making the same select queries
+        on related objects repeatedly
+        """
+        car2 = CarFactory.create()
+        self.driver.cars.add(car2)
+        reset_queries()
+
+        for i in range(5):
+            Driver.objects.get(id=self.driver.id).cars.all()[0].year
+        # The above query is actually composed of 2 sql queries:
+        # one for fetching drivers and the other cars.
+        self.assertEqual(len(connection.queries), 2)
+
+    @override_settings(DEBUG=True)
+    def test_many_to_many_mapping_cache_with_save(self):
+        """
+        Cache should be invalidated when calling 'save' on related objects
+        """
+        car2 = CarFactory.create()
+        self.driver.cars.add(car2)
+        Driver.objects.get(id=self.driver.id).cars.all()[0].year
+        car2.name = 'gti'
+        car2.save()
+        reset_queries()
+
+        # Only 1 cache (the one for car selection query) will be invalidated
+        # as we only update data on Car table
+        Driver.objects.get(id=self.driver.id).cars.all()[0].year
+        self.assertEqual(len(connection.queries), 1)
+
+    @override_settings(DEBUG=True)
+    def test_many_to_many_mapping_cache_with_delete(self):
+        """
+        Cache should be invalidated when calling 'delete' on related objects
+        """
+        car2 = CarFactory.create()
+        self.driver.cars.add(car2)
+        Driver.objects.get(id=self.driver.id).cars.all()[0].year
+        car2.delete()
+        reset_queries()
+
+        # Only 1 cache (the one for car selection query) will be invalidated
```

Comment:
I would have thought there would be 2 queries since a car relation is being deleted